Support ruby 3 and kwargs (latest)

NourEldinShobier opened the pull request:
In this PR we move the changes proposed in the main repo PR https://github.com/collectiveidea/delayed_job/pull/1158 for handling Ruby3 and kwargs

NourEldinShobier committed:
feat: move (Support ruby 3 and kwargs) PR changes

NourEldinShobier committed:
fix: resolve indentations

Commit:
remove ruby 2 support

Commit:
fixing specs, WIP

mokus80 commented on lib/delayed/performable_method.rb:
```diff
@@ -22,18 +23,30 @@ def display_name
+    def kwargs
+      # Default to a hash so that we can handle deserializing jobs that were
+      # created before kwargs was available.
+      @kwargs || {}
```
I don't think we need this any more, do we?

korny replied:
We don't need it anymore, that's right. But it's still important for anyone who switches to this version. So when we create an official Pull Request to the main repo, I would like to include this useful patch.

Can we leave it in to not create two versions?

mokus80 replied:
You want to create a an official PR to https://github.com/collectiveidea/delayed_job? I was not aware of that. In that case we need to keep all the code supporting Ruby versions < 3 of which I have already deleted some. Considering they never paid attention to the [other PR],(https://github.com/collectiveidea/delayed_job/pull/1158) I have doubts that they would consider it.

korny replied:
True. But I think there are other users who are interested in this, and they might choose to use our branch. It's about giving back to the open source community a bit, and practicing best practices for libraries :)

This is a public conversation btw.

mokus80 commented on spec/performable_method_spec.rb:
```diff
@@ -1,9 +1,10 @@
 describe Delayed::PerformableMethod do
   describe 'perform' do
     before do
-      @method = Delayed::PerformableMethod.new('foo', :count, ['o'])
```
@korny @NourEldinShobier I haven't yet figured out why they are passing ['o'] here. Since I don't think it's necessary, I removed the square brackets.

korny replied:
As explained, the brackets are necessary as this needs to be an Array.

Commit:
fixing specs

mokus80 commented on lib/delayed/performable_method.rb:
```diff
@@ -1,8 +1,8 @@
+    attr_accessor :object, :method_name, :args, :kwargs
 
-    def initialize(object, method_name, args)
+    def initialize(object, method_name, args, kwargs = {})
```
@korny @NourEldinShobier We could either keep it this way of requiring args before kwargs. Or we could make args optional, i.e. `def initialize(object, method_name, *args, **kwargs)` - in that case we'll need to change some code.

korny replied:
Not really. The signature should stay as it is for compatibility with older jobs. `args` is an Array and can be empty, but it can not be missing.

Commit:
update specs

Commit:
improve specs

Commit:
more improvements

NourEldinShobier committed:
refactor: add ruby 2 support

NourEldinShobier committed:
refactor: attempt to fix the specs in ruby < 2.5

NourEldinShobier committed:
Revert "refactor: attempt to fix the specs in ruby < 2.5"

This reverts commit ef453f138fc664f21e33269a5b02a0af7a1e272e.

NourEldinShobier committed:
test: resolve ruby 2.4 spec issue

NourEldinShobier commented:
@korny , @mokus80 i did two small fixes that resolved most of ruby 2 failures, can u please check them?

Regarding the remaining errors in ruby < 2.5, i checked the recent merged PRs on the main repo and noticed that they had the same current failures we have. so probably we don't to too

korny commented on Gemfile:
```diff
@@ -69,6 +69,10 @@ group :test do
+group :development, :test do
+  gem 'pry'
+end
+
```
This does not really belong here.

mokus80 replied:
I would like to keep it around as long as needed, will remove it before the merge.

korny commented on spec/helper.rb:
```diff
@@ -25,6 +25,8 @@
 require 'delayed/backend/shared_spec'
 
+require 'pry'
+
```
revert

Commit:
remove ruby 2 support

Commit:
update method missing

Commit:
use super again

Commit:
add kwargs method again

NourEldinShobier committed:
remove all ruby 2 version from ci matrix

NourEldinShobier committed:
Merge remote-tracking branch 'origin/ruby3-and-kwargs-support' into ruby3-and-kwargs-support

NourEldinShobier committed:
remove pry

NourEldinShobier committed:
support jobs without kwargs

We need to deserialize jobs that were created before kwargs was available, so when they are not present, they should be inferred using the final hash argument.

NourEldinShobier committed:
support block arguments in method_missing

Elsioufy committed:
Merge branch 'master' into ruby3-and-kwargs-support
